be/src/exec/pipeline/exchange/sink_buffer.h

```diff
@@ -47,31 +55,72 @@ class SinkBuffer {
     }
 
     ~SinkBuffer() {
-        _pending_chunks.shutdown();
+        _is_finished = true;
+        for (auto* cache : _caches) {
+            delete cache;
+        }
         _thread.join();
-        if (_closure->unref()) {
-            delete _closure;
+        for (auto* closure : _closures) {
+            if (closure->unref()) {
+                delete closure;
+            }
         }
     }
 
-    void add_request(const TransmitChunkInfo& request) { _pending_chunks.put(request); }
+    void add_request(const TransmitChunkInfo& request) {
+        std::lock_guard<std::mutex> l(_mutex);
+        _caches[request.channel_id]->push(request);
+        _cache_empty_cv.notify_one();
+    }
 
     void process() {
         try {
-            while (true) {
-                if (_closure->has_in_flight_rpc()) {
-                    std::this_thread::sleep_for(std::chrono::milliseconds(1));
-                    continue;
+            while (!_is_finished) {
+                {
+                    std::unique_lock<std::mutex> l(_mutex);
+                    bool is_cache_empty = true;
+                    for (auto* cache : _caches) {
+                        if (!cache->empty()) {
+                            is_cache_empty = false;
+                            break;
+                        }
+                    }
+                    if (is_cache_empty) {
+                        _cache_empty_cv.wait(l);
+                    }
                 }
 
-                TransmitChunkInfo info;
-                if (!_pending_chunks.blocking_get(&info)) {
-                    break;
+                const size_t spin_threshould = 1000;
+                size_t spin_iter = 0;
+
+                for (; spin_iter < spin_threshould; ++spin_iter) {
+                    bool find_any = false;
+                    for (auto* cache : _caches) {
+                        if (cache->empty()) {
+                            continue;
+                        }
+
+                        // std::queue' read is concurrent safe without mutex
+                        if (!_closures[cache->front().channel_id]->has_in_flight_rpc()) {
+                            TransmitChunkInfo info = cache->front();
+                            find_any = true;
+                            _send_rpc(info);
+                            {
+                                std::lock_guard<std::mutex> l(_mutex);
+                                cache->pop();
+                            }
+                            // The original design is bad, we must release_finst_id here!
+                            info.params.release_finst_id();
+                        }
+                    }
+
+                    if (find_any) {
```

Comment:
1. index of _caches just is channel id, it's not required to peek front of the queue.
2. here queue.front() is not safe, if someone adds request to the chunk and extends the size of queue, the original memory room occupie by the queue would be released, so it's possible that accessing invalid memory happens when invoke queue.front().
```
for (auto* cache : _caches) {
                        if (cache->empty()) {
                            continue;
                        }

                        // std::queue' read is concurrent safe without mutex
                        if (!_closures[cache->front().channel_id]->has_in_flight_rpc()) {
                            TransmitChunkInfo info = cache->front();
                            find_any = true;
                            _send_rpc(info);
                            {
                                std::lock_guard<std::mutex> l(_mutex);
                                cache->pop();
                            }
                            // The original design is bad, we must release_finst_id here!
                            info.params.release_finst_id();
                        }
                    }
```

Reply:
The container of queue is list not vector, it has no no scaling operation, so it's safe that just read the front of the list